File: vbox-app/src/com/kedzie/vbox/machine/MachineFragment.java
package com.kedzie.vbox.machine;


import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.actionbarsherlock.app.SherlockFragment;
import com.actionbarsherlock.view.Menu;
import com.actionbarsherlock.view.MenuInflater;
import com.actionbarsherlock.view.MenuItem;
import com.kedzie.vbox.R;
import com.kedzie.vbox.SettingsActivity;
import com.kedzie.vbox.app.BundleBuilder;
import com.kedzie.vbox.app.FragmentElement;
import com.kedzie.vbox.app.PagerTabHost;
import com.kedzie.vbox.app.Utils;
import com.kedzie.vbox.soap.VBoxSvc;
import com.kedzie.vbox.task.ConfigureMetricsTask;

public class MachineFragment extends SherlockFragment {
    private static final int REQUEST_CODE_PREFERENCES = 6;

    private PagerTabHost mTabHost;
    private boolean mDualPane;
    private VBoxSvc _vmgr;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mDualPane = getArguments().getBoolean("dualPane");
        _vmgr = BundleBuilder.getVBoxSvc(getArguments());
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mTabHost = new PagerTabHost(getActivity());
        mTabHost.setup(getActivity(), getChildFragmentManager(), R.id.realtabcontent);
        Bundle args =  getArguments();
        args.putBoolean("dualPane", mDualPane);
        mTabHost.addTab(new FragmentElement("Details", R.drawable.ic_menu_info_details, InfoFragment.class, args));
        mTabHost.addTab(new FragmentElement("Actions", ActionsFragment.class, args));
        mTabHost.addTab(new FragmentElement("Log", LogFragment.class, args));
        mTabHost.addTab(new FragmentElement("Snapshots", R.drawable.ic_menu_camera, SnapshotFragment.class, args));
        return mTabHost;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
		if(!mDualPane)
        	inflater.inflate(R.menu.machine_actions, menu);
		((SherlockFragment)mTabHost.getCurrentFragment()).onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()) {
            case R.id.option_menu_preferences:
                Utils.startActivityForResult(getActivity(), new Intent(getActivity(), SettingsActivity.class), REQUEST_CODE_PREFERENCES);
                return true;
        }
		return ((SherlockFragment)mTabHost.getCurrentFragment()).onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(requestCode==REQUEST_CODE_PREFERENCES) {
            new ConfigureMetricsTask(getSherlockActivity(), _vmgr).execute(
                    Utils.getIntPreference(getActivity(), SettingsActivity.PREF_PERIOD),
                    Utils.getIntPreference(getActivity(), SettingsActivity.PREF_COUNT) );
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mTabHost = null;
    }
}

File: vbox-app/src/main/java/com/kedzie/vbox/machine/MachineFragment.kt
package com.kedzie.vbox.machine


import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentPagerAdapter
import androidx.lifecycle.Observer
import androidx.lifecycle.viewModelScope
import com.kedzie.vbox.R
import com.kedzie.vbox.VBoxApplication
import com.kedzie.vbox.api.IMachine
import com.kedzie.vbox.api.IMachineStateChangedEvent
import com.kedzie.vbox.soap.VBoxSvc
import kotlinx.android.synthetic.main.machine_fragment.*
import kotlinx.android.synthetic.main.machine_view.*
import kotlinx.coroutines.launch
import org.koin.androidx.viewmodel.ext.android.sharedViewModel
import org.koin.core.parameter.parametersOf
import timber.log.Timber

class MachineFragment : Fragment() {

    private val model: MachineListViewModel by sharedViewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.machine_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        pager.adapter = object : FragmentPagerAdapter(childFragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {
            override fun getCount() = 4

            override fun getItem(position: Int): Fragment {
                return when (position) {
                    0 -> InfoFragment(arguments!!)
                    1 -> ActionsFragment(arguments!!)
                    2 -> LogFragment(arguments!!)
                    else -> SnapshotFragment(arguments!!)
                }
            }

            override fun getPageTitle(position: Int): CharSequence {
                return when (position) {
                    0 -> getString(R.string.machine_tab_info)
                    1 -> getString(R.string.machine_tab_actions)
                    2 -> getString(R.string.machine_tab_log)
                    else -> getString(R.string.machine_tab_snapshots)
                }
            }
        }
        pager_tabs.setupWithViewPager(pager)

        model.machine.observe(viewLifecycleOwner, Observer {
            machine_view.contentDescription = "Virtual Machine: " + it.getNameNow()
            machine_list_item_ostype.setImageResource(VBoxApplication.getOSDrawable(context, m.getOSTypeId()))
            machine_list_item_name.text = m.getName()
            machine_list_item_state.setImageResource(m.getStateNoCache().drawable())
            machine_list_item_state_text.text = m.getState().value()
            machine_list_item_snapshot.text = m.getCurrentSnapshotNoCache()?.let {
                StringBuffer("(").append(it.getName()).append(")").append(if (m.getCurrentStateModifiedNoCache()) "*" else "").toString()
            } ?: ""
        })
    }
}
